### DISCUSSION TITLE
✅ feature suggestion: websocket support

### DISCUSSION THREAD

**Discussion by uxduck** [RobinTail/express-zod-api#991] [Ideas]
It would be nice if we could define WebSocket endpoints that clients can call, which establishes a connection. The handler would give developers a `subscription`, which they can store to send messages to the client later, or close the websocket.

## Example
- Multiple users call the `subscribeEndpoint`, which sets up a websocket between them and the server. 
- Whenever a user calls the `sendMessageEndpoint`, messages are broadcast to all subscribers.
- Clients can unsubscribe (client side code not shown)

```typescript
// In practice, subscriptions might be scoped to a specific chat room, not globally for all users.
const subscriptions: Subscription[] = [];

const subscribeEndpoint = defaultEndpointsFactory.build({
    method: "subscribe",
    input: z.object({
      userId: z.string(),
    }),
    message: z.object({ // defines the message type on the websocket
      message: z.string(),
      userId: z.string(),
    }),
    output: z.void({}),
    handler: async ({ input: { name }, options, logger, subscription }) => {
      subscriptions.push(subscription);
        logger.debug(`Subscribed by ${input.userId}`, options); 
      return {};
    },
  });

const sendMessageEndpoint = defaultEndpointsFactory.build({
    method: "post",
    input: z.object({
        userId: z.string(),
        message: z.string()
      }),
    output: z.void(),
    handler: async ({ input: { name }, options, logger, subscription }) => {
        for (const sub of subscriptions) {
            // send message to all subscribers
            sub.sendMessage({ message: input.message, userId: input.userId });
        };
      },
});
```

**Comment by RobinTail** [comment]
Hello @ben-xD ,

I'm not sure I understand what "WebSocket endpoints" are.
As far as I know websockets are event-based entities operating own protocol, but maybe something changed since last time I worked with websockets.

From your code samples, if I understand correctly, you meant that there are similarities between Request `method` and Event type, and that the Event payload could be validated against the schema.

My experience with Websockets is limited to this library:
https://socket.io/get-started/chat#integrating-socketio

And as you may see from the example there, it has a slightly separated workflow (dedicated event-based server).

So... I'm not sure that the existing entities (like EndpointsFactory) could be reused in any suitable way for integrating websockets.

However, since `zod` is a peer dependency now, and following `attachRouting()` approach with a custom app, it could be actually not a difficult task to add websockets with IO validation by consumer, I presume.

On the other side, making websockets entry easier seems interesting.

Please let me know if you have any recommendations or visions on how this idea could be implemented.
I'll take a time to think on it as well. Thank you.

**Comment by RobinTail** [comment]
Dear @ben-xD ,

would you be interested in testing this feature:
https://github.com/RobinTail/express-zod-api/pull/1545

Could it help you to achieve the desired functionality?

Your feedback is appreciated — please let me know.

**Reply by uxduck** [reply]
Hey @RobinTail

Sorry for not replying earlier. I've been using trpc over websockets which has been quite nice. I could play around with your PR/feature though :)

**Comment by RobinTail** [comment]
Hello @ben-xD ,

I actually decided to move that feature to a separate project, which is basically attachable to any HTTP server, including the one returned by `createServer()` of `express-zod-api`.

Check this out: https://www.npmjs.com/package/zod-sockets

Would you try to install version 0.2.0 or higher for your sample project and let me know if it's usable, understandable, make sense, etc.
Basically, I need some feedback.

So far, this thing is capable to listen for incoming events, acknowledge them, emit and broadcast events and handle their acknowledgements. All that is validated and you can connect the logger from `express-zod-api` to it.
There is an example folder in [the repo](https://github.com/RobinTail/zod-sockets) where I managed to demonstrate such samples as ping-pong with echo, subscriptions and a basic chat.

Rooms and namespaces are not yet supported.
I've been using [Postman](https://learning.postman.com/docs/sending-requests/websocket/create-a-socketio-request/) for playing with it.
Please let me know what you think about it.

**Comment by RobinTail** [comment]
Latest news on that matter for those who interested, CC @ben-xD :

https://github.com/RobinTail/zod-sockets

- I significantly improved a lot of things there;
- Made a very simple example of subscription service;
- It has all the basic features including rooms and namespaces;
- It can generate documentation according to AsyncAPI standard;
- It can export types for client constraints;
- It's awesome!

Check it out.

**Comment by RobinTail** [comment]
Dear @ben-xD ,

did you have a chance to look at or try [zod-sockets](https://github.com/RobinTail/zod-sockets) ?
I need some feedback.

**Comment by RobinTail** [comment]
I made a lightweight alternative using Server-sent Events.
Client can subscribe to event stream. No extra package needed.
Available in v21.5.0, CC @ben-xD 
https://ez.robintail.cz/v21.5.0/subscriptions

**Reply by jakub-msqt** [reply]
I followed the example from the docs and I have 2 questions:

1. How can I close the connection from the server?
2. I don't want to use `setTimeout` but rather wait for an event to happen, e.g. notification from redis. With the `while` loop, it's executing the logic inside over and over again. What's the solution here?

**Reply by RobinTail** [reply]
> How can I close the connection from the server?

just exit the loop, @jakub-msqt 

> I don't want to use setTimeout but rather wait for an event to happen, e.g. notification from redis. With the while loop, it's executing the logic inside over and over again. What's the solution here?

The loop is required to keep the connection open;
The `setTimeout()` is an example of implementing a delay between sending the messages, you can `await` for receiving an event from Redis. For that, I presume, you'd need to utilize the Redis connection instance within the `handler`. You can either declare it in a higher order scope OR supply it via `options` using the `.addOptions()` or `.addMiddleware()` (be careful with this approach since it executes for every connection).

**Reply by jakub-msqt** [reply]
I think I'm almost there with the AsyncIterator approach. The only issue is that I don't see the exit log on the server when client closes connection.

```ts
subscriptionEndpointsFactory.buildVoid({
  handler: async ({ input, options: { emit, isClosed } }) => {
    const ee = new EventEmitter()
    const eventName = 'update'

    emit('command', 'WELCOME')

    setTimeout(() => {
      ee.emit(eventName, 'DELAYED EVENT')
    }, 3000)

    for await (const [payload] of on(ee, eventName)) {
      if (isClosed()) {
        break
      }

      emit('command', payload)
    }

    console.log('EXIT')
  },
})
```

tRPC [solves](https://trpc.io/docs/server/subscriptions#basic-example) this with `AbortSignal` passed to `on` as `on(ee, eventName, { signal })`. Could something like that be added? Or is there a different solution?

**Reply by RobinTail** [reply]
can't figure out: what is `on()`, @jakub-msqt ?

**Reply by RobinTail** [reply]
Ah!

```ts
import EventEmitter, { on } from 'events';
```

**Reply by RobinTail** [reply]
> tRPC [solves](https://trpc.io/docs/server/subscriptions#basic-example) this with AbortSignal passed to on as on(ee, eventName, { signal }). Could something like that be added? Or is there a different solution?

Sure, here is the place where I think it could be added:
https://github.com/RobinTail/express-zod-api/blob/cc3d6b66b12d0e1ee5108349728545e52b4fd092/express-zod-api/src/sse.ts#L56-L71

I welcome your PR, @jakub-msqt

**Reply by jakub-msqt** [reply]
Ok, I'll give it a go.

**Reply by jakub-msqt** [reply]
https://github.com/RobinTail/express-zod-api/pull/2965
